### PR TITLE
Fix: Prevent RELIABLE_RESET offset increase by using correct sentinel

### PR DESCRIPTION
**Pull request by basavaraj-sm05**
## Description
Fixes a bug where a peer could bypass the strictly-decreasing offset rule for `RELIABLE_RESET` frames.
Right now, `QuicStreamProcessReliableResetFrame` checks `Stream->RecvMaxLength == 0` to see if this is the first `RELIABLE_RESET` we've received. But `0` is actually a valid `ReliableOffset` value! 
If a peer sends a `RELIABLE_RESET` with offset `0`, we update `RecvMaxLength` to `0`. If they then send another one with offset `5000`, the `== 0` check still passes, and we end up increasing the offset to `5000` — which violates the spec.
So instead of checking against `0`, this PR just uses the `RemoteCloseResetReliable` flag to check if we've seen a reliable reset frame yet.
## Testing
Existing reliable reset tests pass fine since standard behavior isn't affected. 
We might want to add a quick test case that sends a `RELIABLE_RESET(offset=0)` followed by a `RELIABLE_RESET(offset=5000)` to verify it gets rejected properly now.
## Documentation
None needed, just an internal state fix.

### CODE REVIEW

File: src/core/stream_recv.c [hunk 1]
```diff
@@ -206,10 +206,12 @@ QuicStreamProcessReliableResetFrame(
...
 
-    if (Stream->RecvMaxLength == 0 || ReliableOffset < Stream->RecvMaxLength) {
+    if (!Stream->Flags.RemoteCloseResetReliable || ReliableOffset < Stream->RecvMaxLength) {
         //
-        // As outlined in the spec, if we receive multiple CLOSE_STREAM frames, we only accept strictly
-        // decreasing offsets.
+        // As outlined in the spec, if we receive multiple RELIABLE_RESET frames, we only accept strictly
+        // decreasing offsets. Use the RemoteCloseResetReliable flag (not RecvMaxLength == 0) as the
+        // "first frame seen" sentinel because ReliableOffset = 0 is a valid value and would otherwise
+        // allow a subsequent frame with a larger offset to bypass this check.
```
Comment (guhetier) [inline src/core/stream_recv.c]:
```suggestion
        // If multiple RELIABLE_RESET frames are received, only accept strictly
        // decreasing offsets.
```

Let's keep the comment to the point.

**Review by guhetier — COMMENTED** [review]
Thanks for the PR.
Some nitpick on the comment, but looks good to me otherwise. 
Adding a test for this corner case would be ideal.

File: src/core/stream_recv.c (hunk 1 above)
Comment (basavaraj-sm05) [reply src/core/stream_recv.c]: Thanks for the review @guhetier!

I've pushed an update addressing both points

**Comment by basavaraj-sm05** [comment]
@microsoft-github-policy-service agree company="Digiscrypt Technologies"

File: src/test/lib/DataTest.cpp [hunk 2]
```diff
@@ -4521,6 +4567,111 @@ QuicTestStreamReliableResetMultipleSends(
...
+        TEST_QUIC_SUCCEEDED(Stream.Start());
+
+        //
+        // Send 1 byte so QueuedSendOffset >= 1, making SetReliableOffset(0) valid.
+        //
+        TEST_QUIC_SUCCEEDED(Stream.Send(&SendBuffer, 1, QUIC_SEND_FLAG_DELAY_SEND, nullptr));
+
+        //
+        // Set ReliableOffset to 0 — this is the corner case. The old code used
+        // RecvMaxLength == 0 as "first frame not yet received", which means
```
Comment (guhetier) [inline src/test/lib/DataTest.cpp]: Let's avoid comments referring to the past state of the code, that won't be relevant for future readers.
As far as they are concerned, this test validates the corner case where the reliable offset is 0.

File: src/test/lib/DataTest.cpp [hunk 3]
```diff
@@ -4521,6 +4567,111 @@ QuicTestStreamReliableResetMultipleSends(
...
+
+    MsQuicConnection Connection(Registration);
+    TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+
+    TEST_QUIC_SUCCEEDED(Connection.Start(ClientConfiguration, ServerLocalAddr.GetFamily(), QUIC_TEST_LOOPBACK_FOR_AF(ServerLocalAddr.GetFamily()), ServerLocalAddr.GetPort()));
+    TEST_TRUE(Connection.HandshakeCompleteEvent.WaitTimeout(TestWaitTimeout));
+    TEST_TRUE(Connection.HandshakeComplete);
+    TEST_TRUE(Listener.LastConnection->HandshakeCompleteEvent.WaitTimeout(TestWaitTimeout));
+    TEST_TRUE(Listener.LastConnection->HandshakeComplete);
+    CxPlatSleep(50); // Wait for things to idle out
```
Comment (guhetier) [inline src/test/lib/DataTest.cpp]: I don't think there is a reason to wait for this test?

File: src/test/lib/DataTest.cpp [hunk 4]
```diff
@@ -4521,6 +4567,111 @@ QuicTestStreamReliableResetMultipleSends(
...
+    MsQuicConfiguration ServerConfiguration(Registration, "MsQuicTest", TestSettings, ServerSelfSignedCredConfig);
+    TEST_QUIC_SUCCEEDED(ServerConfiguration.GetInitStatus());
+
+    MsQuicConfiguration ClientConfiguration(Registration, "MsQuicTest", TestSettings, MsQuicCredentialConfig());
+    TEST_QUIC_SUCCEEDED(ClientConfiguration.GetInitStatus());
+
+    StreamReliableResetZeroOffset Context;
+
+    //
+    // We need to send at least 1 byte so that QueuedSendOffset > 0 and
```
Comment (guhetier) [inline src/test/lib/DataTest.cpp]: nit: for the sake of it, I'd send more than 1 byte, so there is data after the reliable reset offset.

File: src/test/lib/DataTest.cpp [hunk 5]
```diff
@@ -4521,6 +4567,111 @@ QuicTestStreamReliableResetMultipleSends(
...
+// spec's strictly-decreasing rule.
+//
+// The fix uses the RemoteCloseResetReliable flag as the sentinel instead.
+// This test confirms that after receiving RELIABLE_RESET(offset=0), the
+// server correctly tracks the received offset as 0 (i.e.
+// QUIC_PARAM_STREAM_RELIABLE_OFFSET_RECV returns 0, not INVALID_STATE).
+// If the old sentinel were still in place, the flag would still be set to
+// TRUE and RecvMaxLength would be 0, so RecvMaxLength == 0 would wrongly
+// look like "no frame received yet" — leaving the door open for a higher-
+// offset frame to pass the check.
```
Comment (guhetier) [inline src/test/lib/DataTest.cpp]: Let's remove this or shorten it to a simple description of what the test validates without reference to the past state of the code (e.g. "validates that a reliable-reset of `0` is handled properly").

File: src/core/sliding_window_extremum.c
Comment (guhetier) [inline src/core/sliding_window_extremum.c:1]: Changes to this file don't seem to belong with this PR.

File: src/test/lib/DataTest.cpp (hunk 2 above)
Comment (basavaraj-sm05) [reply src/test/lib/DataTest.cpp]: Fair enough! Removed the references to the old behavior and kept it focused on the corner case itself.

File: src/test/lib/DataTest.cpp (hunk 3 above)
Comment (basavaraj-sm05) [reply src/test/lib/DataTest.cpp]: You're right, there's no real need for it here. Removed!

File: src/core/sliding_window_extremum.c
Comment (basavaraj-sm05) [reply src/core/sliding_window_extremum.c:1]: Oops, sorry about that! That was an unrelated fix I had locally and it accidentally got bundled in. Reverted it — it's out of this PR now.

File: src/test/lib/DataTest.cpp (hunk 4 above)
Comment (basavaraj-sm05) [reply src/test/lib/DataTest.cpp]: Good nit! Bumped it to 16 bytes so there's actually some data sitting after the reliable reset offset, which makes the test feel more realistic.

File: src/test/lib/DataTest.cpp (hunk 5 above)
Comment (basavaraj-sm05) [reply src/test/lib/DataTest.cpp]: Shortened it in the last push, it's down to a brief description of what the test validates now.

**Comment by basavaraj-sm05** [comment]
any update?

**Review by guhetier — DISMISSED** [review]

File: src/test/lib/DataTest.cpp [hunk 6]
```diff
@@ -4364,6 +4364,52 @@ struct StreamReliableReset {
...
+
+    static QUIC_STATUS ConnCallback(_In_ MsQuicConnection*, _In_opt_ void* Context, _Inout_ QUIC_CONNECTION_EVENT* Event) {
+        auto TestContext = (StreamReliableResetZeroOffset*)Context;
+        if (Event->Type == QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED) {
+            TestContext->ServerStream =
+                new(std::nothrow) MsQuicStream(
+                    Event->PEER_STREAM_STARTED.Stream,
+                    CleanUpAutoDelete,
+                    ServerStreamCallback,
+                    Context);
```
[the remark this answers is not in the document]
Comment (basavaraj-sm05) [reply src/test/lib/DataTest.cpp:4406]: Good catch. Fixed in 5a61500: the server stream wrapper is created with CleanUpManual now, and the test deletes it explicitly after querying the reliable offset, so it can't free itself on shutdown-complete before the GetReliableOffsetRecv call.
